src/libcore/iter/traits/iterator.rs

```diff
@@ -1992,17 +2001,16 @@ pub trait Iterator {
     /// ```
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
-    fn fold<B, F>(mut self, init: B, f: F) -> B
+    fn fold<B, F>(mut self, init: B, mut f: F) -> B
     where
         Self: Sized,
         F: FnMut(B, Self::Item) -> B,
     {
-        #[inline]
-        fn ok<B, T>(mut f: impl FnMut(B, T) -> B) -> impl FnMut(B, T) -> Result<B, !> {
```

Comment:
Maybe this is totally beside the point, but is the type `Result<B, !>` the best we can do in terms of reducing instantiation noise in infallible `try_fold`? Is it likely that something could be gained with an even simpler `#[repr(transparent)] struct Value<T>(T)`  type that implements `Try`?

I guess that having a newtype wrapper at all, be it `Ok::<T, !>` or `Value` is instantiation noise enough, combined with the conversion to Result by the try operator.

But if we could use specialization to remove the use of `?` entirely in `try_fold` for certain types? (I'm only making a bigger mess, am I not?)

Reply:
I don't have an opinion on this, other than it seems like something for a follow-up, given that it's a pre-existing code pattern than is orthogonal to this PR :)